design-proposals/klip-42-schema-migrations-tool.md

```diff
@@ -0,0 +1,442 @@
+
+For instance, let's undo the migration version 2 applied before. The migration file used before was named `V2__Add_users.sql`. For the undo file, we should add the `U` prefix.
+
+`U2_Add_users.sql`
+```sql
+DROP TABLE users;
+```
+
+The user runs the migration tool on the ksqlDB cluster. The tool detects the cluster has version 2, so it executes only the undo file for version 2.  
+It then sets the state of the cluster to version 1.
+
+```shell script
+$ ksql-migrations undo
+Current version of schema: 2
+Undoing migration of schema to version 2 - Add users
+```
+
+The `undo` action will only revert the previous change. It will not attempt to undo all applied migrations.
+
+
+### Naming convention
+
+Hope you have noticed the naming rules I followed in the previous examples for naming the migration files. Having a naming convention for files is necessary
+for the tool to detect what migration to apply or revert. Naming files are easier than creating configurations or command parameters to specify the same info.
+
+The migration files follow the same `Flyway` naming convention.
+`(Prefix)(Version)__(Description).sql`
+
+The `Prefix` specifies whether the file is a new migration (`V`) or undo (`U`) file.
+The `Version` is the version number used for the schema. For minor versions, such as `1.1`, an underscore is required (i.e. `1_1`)
+The `Description` is just a name or description of the new migration.
+
+i.e.
+```
+- V1__Initial_setup.sql    # a new version to migrate (v1)
+- U1__Initial_setup.sql    # rollback v1 schema
+- V2__Add_users.sql        # a new version to migrate (v2)
+- U2__Add_users.sql        # rollback v2 schema
+- V2_1__Fix_topic_name.sql # A new version to migrate (v2.1)
+- U2_1__Fix_topic_name.sql # rollback v2.1 schema
+```
+
+It is recommended the user creates these two files on any new migration. The tool will not enforce that. We need specify this recomendation in the ksqlDB documents.
+
+### Command Syntax
+
+Finally, let's look at the rest of the command parameters that will exist to facilitate schema migrations.
+
+```shell script
+Usage:
+  ksql-migrations [options] commands
+  
+Commands
+  initialize   Initializes the schema version table
+
+  apply ( all | next | until <target> )
+  
+              Migrates a schema to new available schema versions (default: next)
+              
+              If 'all' is specified, then it migrates all newer versions available
+              If 'next' is specified, then it migrates only the next available version
+              If 'until <target>' is specified, then it migrates all available versions before <target>
+  
+  undo ( all | last | until <target> )
+  
+              Rollbacks a schema to the previous schema version (default: last)
+              
+              If 'all' is specified, then it rollbacks all previous versions
+              If 'last' is specified, then it rollbacks the previous version
```

Comment:
Can we consider `previous` instead? I think this mirrors apply's `next` option better, personally.

Reply:
Done